producers/http/http.go

```diff
@@ -0,0 +1,106 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package http
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/coreos/go-systemd/activation"
+	"github.com/dcos/dcos-go/store"
+	"github.com/dcos/dcos-metrics/producers"
+)
+
+// Config for the HTTP producer
+type Config struct {
+	Port        int           `yaml:"port"`
+	CacheExpiry time.Duration // ideally this is a multiple of the collector's PollingPeriod
+}
+
+type producerImpl struct {
+	config      Config
+	store       store.Store
+	metricsChan chan producers.MetricsMessage
+}
+
+// New creates a new instance of the HTTP producer with the provided configuration.
+func New(cfg Config) (producers.MetricsProducer, chan producers.MetricsMessage) {
+	p := producerImpl{
+		config:      cfg,
+		store:       store.New(),
+		metricsChan: make(chan producers.MetricsMessage),
+	}
+	return &p, p.metricsChan
+}
+
+// Run a HTTP server and serve the various metrics API endpoints.
+// This function should be run in its own goroutine.
+func (p *producerImpl) Run() error {
+	log.Info("Starting HTTP producer garbage collection service")
+	go p.janitor()
+
+	go func() {
+		log.Debug("HTTP producer listening for incoming messages on metricsChan")
+		for {
+			message := <-p.metricsChan // read messages off the channel
+			log.Debugf("Received message '%s' with timestamp %s",
+				message.Name,
+				time.Unix(message.Timestamp, 0).Format(time.RFC3339))
+			log.Debugf("Setting store object '%s' with timestamp %s",
+				message.Name,
+				time.Unix(message.Timestamp, 0).Format(time.RFC3339))
+			p.store.Set(message.Name, message) // overwrite existing object with the same message.Name
+		}
+	}()
+
+	r := newRouter(p)
+	listeners, err := activation.Listeners(true)
+	if err != nil {
+		return fmt.Errorf("Unable to get listeners: %s", err)
+	}
+	// If a listener is avialable, use that. If it is not avialable,
+	// listen on the default TCP socket and port.
+	if len(listeners) == 1 {
+		log.Infof("HTTP Producer serving requests on %s", listeners[0].Addr().String())
+		return http.Serve(listeners[0], r)
+	} else {
+		log.Infof("HTTP Producer serving requests on :%d", p.config.Port)
+		return http.ListenAndServe(fmt.Sprintf(":%d", p.config.Port), r)
+	}
+}
+
+// janitor analyzes the objects in the store and removes stale objects. An
+// object is considered stale when the top-level timestamp of its MetricsMessage
+// has exceeded the CacheExpiry, which is calculated as a multiple of the
+// collector's polling period. This function should be run in its own goroutine.
+func (p *producerImpl) janitor() {
```

Comment:
Question: does that functionality belong in dcos-metrics, or dcos-go/store? :thinking: